[css-anchor-position-1] Parse anchor() as a calc() function

anttijk opened the pull request:
#### 9e5342da427ea97dfa23896043fa0bd0b20bb36b
<pre>
[css-anchor-position-1] Parse anchor() as a calc() function
<a href="https://bugs.webkit.org/show_bug.cgi?id=280191">https://bugs.webkit.org/show_bug.cgi?id=280191</a>
<a href="https://rdar.apple.com/136514213">rdar://136514213</a>

Reviewed by Darin Adler and Sam Weinig.

Add calc tree node and parsing and serialization support for anchor().
The new code is only used in the calc(anchor(foo)) case for now, not with a top level anchor().

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-invalid-fallback-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-parse-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-query-custom-property-registration-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-typed-om-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/global-attributes/the-anchor-attribute-003.tentative-expected.txt:

New FAILs are in previously unimplemented feature.

* Source/WebCore/css/calc/CSSCalcTree+CalculationValue.cpp:
(WebCore::CSSCalc::toCalculationValue):
* Source/WebCore/css/calc/CSSCalcTree+ComputedStyleDependencies.cpp:
(WebCore::CSSCalc::collectComputedStyleDependencies):
* Source/WebCore/css/calc/CSSCalcTree+Copy.cpp:
(WebCore::CSSCalc::copy):
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.cpp:
(WebCore::CSSCalc::evaluate):
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
(WebCore::CSSCalc::isCalcFunction):
(WebCore::CSSCalc::consumeAnchor):
(WebCore::CSSCalc::parseCalcFunction):
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
(WebCore::CSSCalc::serializeMathFunctionArguments):
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
(WebCore::CSSCalc::simplify):
(WebCore::CSSCalc::copyAndSimplifyChildren):
* Source/WebCore/css/calc/CSSCalcTree+Simplification.h:
* Source/WebCore/css/calc/CSSCalcTree.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Ident.cpp:
(WebCore::CSSPropertyParserHelpers::consumeDashedIdentRaw):
(WebCore::CSSPropertyParserHelpers::consumeDashedIdent):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Ident.h:

Canonical link: <a href="https://commits.webkit.org/284215@main">https://commits.webkit.org/284215@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/91b04585d634162426a3bdf2b35f9371d536ecf0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68765 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48157 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21424 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72835 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19910 "Built successfully") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/70882 "Build is in progress. Recent messages:OS: Ventura (13.6.7), Xcode: 14.3; Running apply-patch; Checked out pull request; Passed bindings tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55953 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19726 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/54807 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13241 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71832 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43999 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59374 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/35273 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40664 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16800 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18268 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62613 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17148 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74529 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12736 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16392 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/62295 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12776 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59456 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/62331 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15263 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10287 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3900 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43958 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45032 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46226 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44774 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-invalid-fallback-expected.txt:
```diff
@@ -19,8 +19,8 @@ PASS Invalid anchor function,  cross-axis query (horizontal)
 PASS Invalid anchor function, nested left
-PASS Invalid anchor function, nested right
-PASS Invalid anchor function, nested bottom
+FAIL Invalid anchor function, nested right assert_equals: left expected "0px" but got "188.4375px"
+FAIL Invalid anchor function, nested bottom assert_equals: top expected "0px" but got "182px"
```
Will we be fixing this later?

anttijk replied:
Yes. The PASSes were not real, everything was invalid since we weren't parsing the case before.

darinadler commented on LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-parse-valid-expected.txt:
```diff
@@ -2015,7 +2015,7 @@ PASS e.style['inset-inline-end'] = "anchor(--foo 50%, anchor(--bar left))" shoul
-FAIL e.style['top'] = "calc((anchor(--foo top) + anchor(--bar bottom)) / 2)" should set the property value assert_not_equals: property should be set got disallowed value ""
-FAIL e.style['top'] = "anchor(--foo top, calc(anchor(--bar bottom) * 0.5))" should set the property value assert_not_equals: property should be set got disallowed value ""
-FAIL e.style['top'] = "min(100px, 10%, anchor(--foo top), anchor(--bar bottom))" should set the property value assert_not_equals: property should be set got disallowed value ""
+FAIL e.style['top'] = "calc((anchor(--foo top) + anchor(--bar bottom)) / 2)" should set the property value assert_equals: serialization should be canonical expected "calc((anchor(--foo top) + anchor(--bar bottom)) / 2)" but got "calc(0.5 * (anchor(--foo top) + anchor(--bar bottom)))"
+FAIL e.style['top'] = "anchor(--foo top, calc(anchor(--bar bottom) * 0.5))" should set the property value assert_equals: serialization should be canonical expected "anchor(--foo top, calc(anchor(--bar bottom) * 0.5))" but got "anchor(--foo top, calc(0.5 * anchor(--bar bottom)))"
```
These are annoying. Should we be changing the test to allow our result? Or do we need to fix this to be exactly what the test currently expects?

anttijk replied:
That seems to be a general issue with our calc() serialization. Not sure what the spec says, but it is better handled separately.

darinadler commented on LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-query-custom-property-registration-expected.txt:
```diff
@@ -1,5 +1,10 @@
 
-PASS anchor() cannot be used as <length> initial value
+FAIL anchor() cannot be used as <length> initial value assert_throws_dom: function "() => CSS.registerProperty({
+    name: '--x',
+    syntax: '<length>',
+    inherits: false,
+    initialValue: 'anchor(--foo top)',
+  })" did not throw
```
Would it be easy to fix this?

anttijk replied:
It should be fairly trivial but it is better handled separately.

darinadler commented on Source/WebCore/css/calc/CSSCalcTree+CalculationValue.cpp:
```diff
@@ -263,6 +264,12 @@ template<typename Op> Calculation::Child toCalculationValue(const IndirectNode<O
 
+Calculation::Child toCalculationValue(const IndirectNode<Anchor>&, const ToConversionOptions&)
+{
+    ASSERT_NOT_REACHED("Unevaluated anchor() functions are not supported in the Calculation::Tree");
```
Inelegant, but I don’t have a better idea.

weinig replied:
This is what we do with symbols and non-canonical dimension values, so it's in good company.

weinig replied:
That said, it kinda seems like we will need some for representation in the Calculation Tree, as it seems like the grammar allows for percentages in side the anchor function.

For instance:

```css
width: calc(10% + anchor(--x calc(20% / 2));
```  

Does the 20% inside the anchor get resolved by the style builder? Or does it have to wait until there is a container to do the math like other percentages?

anttijk replied:
The percentage in the anchor function refers to the anchor side (left=0%, right=100%, percent values are between) and can be resolved at computed value time. You can have anything in the fallback but we handle that by simply replacing the anchor node with its fallback node during simplification so the anchor node does not show up in the calculation tree in that case either.

darinadler commented on Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
```diff
@@ -1279,6 +1279,20 @@ std::optional<Child> simplify(Sign& root, const SimplificationOptions& options)
+    bool isValid = !anchor.elementName.isNull() || !options.conversionData->style()->positionAnchor().isNull();
+    if (!isValid && anchor.fallback) {
+        // Replace the anchor node with the fallback node.
+        return WTFMove(anchor.fallback);
```
Should this be `std::exchange` instead of `WTFMove`? If anyone ever reads from `anchor.fallback` after this point, then the answer is probably yes, but if all we do with it is destroy the anchor, then move is fine.

darinadler commented on Source/WebCore/css/calc/CSSCalcTree.h:
```diff
@@ -704,6 +707,24 @@ struct Sign {
 
+
+struct Anchor {
+    WTF_MAKE_TZONE_ALLOCATED_INLINE(Anchor);
+public:
```
Isn’t there a version of the macro for `struct` that leaves things `public`? I guess maybe not.

anttijk replied:
I don't see one, feels like there should be.

darinadler commented on Source/WebCore/css/calc/CSSCalcTree.h:
```diff
@@ -704,6 +707,24 @@ struct Sign {
+    AtomString elementName;
+    Side side;
+
+    std::optional<Child> fallback;
```
I think this would be easier to mentally "parse" without the blank line.

weinig commented on Source/WebCore/css/calc/CSSCalcTree+Evaluation.cpp:
```diff
@@ -165,6 +166,19 @@ std::optional<double> evaluate(const IndirectNode<Hypot>& root, const Evaluation
+    if (!options.conversionData || !options.conversionData->style())
+        return { };
+
+    // FIMXE: Evaluate the anchor.
```
Typo: FIMXE -> FIXME.

weinig commented on Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
```diff
@@ -1279,6 +1279,20 @@ std::optional<Child> simplify(Sign& root, const SimplificationOptions& options)
+    if (!options.conversionData || !options.conversionData->style())
+        return { };
+
+    // FIMXE: Evaluate the anchor.
```
Typo: FIMXE -> FIXME.

anttijk replied:
anchor-calc [~/webkit/OpenSource] git grep FIMXE Source | wc
      13     155    1590

🏛️

weinig replied:
Well I'll be!

weinig commented on Source/WebCore/css/calc/CSSCalcTree.h:
```diff
@@ -1148,6 +1169,7 @@ OP_TUPLE_LIKE_CONFORMANCE(Log, 2);
 OP_TUPLE_LIKE_CONFORMANCE(Exp, 1);
 OP_TUPLE_LIKE_CONFORMANCE(Abs, 1);
 OP_TUPLE_LIKE_CONFORMANCE(Sign, 1);
+OP_TUPLE_LIKE_CONFORMANCE(Anchor, 0);
```
Would be good to implement this with 3, and then have a proper get() implementation, but obviously that has bunch of fallout that would need to be resolved.

The main things you are missing without this are:

- support for collectComputedStyleDependencies() (in CSSCalcTree+ComputedStyleDependencies)
- support for copy() (in CSSCalcTree+Copy)

You can manually add overloads for Anchor in the meantime.

anttijk replied:
I tried that initially but it was just too much template-fu for this patch. This is the first case where children are something other than Nodes.

weinig replied:
Yeah, I need to make that easier. I have an idea I will try out when I implement random.

anttijk committed:
[css-anchor-position-1] Parse anchor() as a calc() function
https://bugs.webkit.org/show_bug.cgi?id=280191
rdar://136514213

Reviewed by Darin Adler and Sam Weinig.

Add calc tree node and parsing and serialization support for anchor().
The new code is only used in the calc(anchor(foo)) case for now, not with a top level anchor().

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-invalid-fallback-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-parse-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-query-custom-property-registration-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/anchor-typed-om-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/dom/elements/global-attributes/the-anchor-attribute-003.tentative-expected.txt:

New FAILs are in previously unimplemented feature.

* Source/WebCore/css/calc/CSSCalcTree+CalculationValue.cpp:
(WebCore::CSSCalc::toCalculationValue):
* Source/WebCore/css/calc/CSSCalcTree+ComputedStyleDependencies.cpp:
(WebCore::CSSCalc::collectComputedStyleDependencies):
* Source/WebCore/css/calc/CSSCalcTree+Copy.cpp:
(WebCore::CSSCalc::copy):
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.cpp:
(WebCore::CSSCalc::evaluate):
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
(WebCore::CSSCalc::isCalcFunction):
(WebCore::CSSCalc::consumeAnchor):
(WebCore::CSSCalc::parseCalcFunction):
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
(WebCore::CSSCalc::serializeMathFunctionArguments):
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
(WebCore::CSSCalc::simplify):
(WebCore::CSSCalc::copyAndSimplifyChildren):
* Source/WebCore/css/calc/CSSCalcTree+Simplification.h:
* Source/WebCore/css/calc/CSSCalcTree.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Ident.cpp:
(WebCore::CSSPropertyParserHelpers::consumeDashedIdentRaw):
(WebCore::CSSPropertyParserHelpers::consumeDashedIdent):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Ident.h:

Canonical link: https://commits.webkit.org/284215@main
